### DISCUSSION TITLE
HOW TO - Set insert as the default in ALL applications, not just on the command line?

### DISCUSSION THREAD

**Discussion by jbrown123** [chrisant996/clink#677] [Q&A] [answered]
The default in the command line with clink is "insert mode" (vs. overwrite). However, in CLI applications (notably SQLITE3) the default is OVERWRITE which is super annoying.

In the old version (0.4.9) I had it set so insert was always the default in every app and the command line.

How do I do this in the new version?

**Comment by chrisant996** [comment, marked as answer]
### Clearing up some misunderstandings:

> The default in the command line with clink is "insert mode" (vs. overwrite).

The default default in Windows consoles is "insert mode".

I say "default default" because in LNK files you can override the default in the console session started from the LNK file, resulting in a per-console custom default.

> However, in CLI applications (notably SQLITE3) the default is OVERWRITE which is super annoying.

CLI applications start out by inheriting the current mode.
But they can call SetConsoleMode to set the mode to use "insert mode" or "overwrite mode".

Clink cannot prevent CLI applications from calling SetConsoleMode and forcing the OS to use "insert mode" or "overwrite mode".

> In the old version (0.4.9) I had it set so insert was always the default in every app and the command line.

Clink v0.4.9 always saves the current console mode before letting you edit the input line, and then restores the console mode to the saved mode after you finish editing the input line:  [refer to the source code here](https://github.com/chrisant996/clink/blob/2fd2c2bbad3820c5e4e39b200a9b6e5d893ddf54/clink/dll/shell_cmd.c#L289).

So, no, you didn't configure _Clink_ 0.4.9 to force every app to use insert mode.  That isn't possible.
But you might have configured something else, such as the Options in the console Properties page (which isn't related to Clink).

### Ok then what is it?

I don't know what's doing it.  But it's not Clink, and never was.

It could be that SQLite is doing it.  You could examine the source code for how it uses SetConsoleMode.

It could be that the console you're using is configured to use overwrite mode by default.  LNK files have a setting in the Properties | Options page that controls whether the console starts out in "insert mode".

> [!NOTE]
> If you open the Properties page for an already-running console, then it shows the _current_ options, not the _startup_ options.  So if you open the Properties page in an already-running console while it's waiting for an input line, then that won't reveal what the mode was before Clink intercepted the input line read request.
>
> To find the console's current input mode _outside of Clink_, you would need to open the Properties page while Clink is not accepting an input line.  One way to accomplish that could be to run `dir /s c:\` and then while it's spewing output, right click in the console title bar and choose Properties from the context menu.  But I'm assuming your using the legacy conhost.  If you're using some other terminal program, then you'd have to consult the other terminal program's documentation to find how to check its input mode.

**Reply by jbrown123** [reply]
That (configuring the console properties to default to insert mode) indeed fixed it.

I upgraded from 0.4.9 to 1.6.18 and this behavior appeared. I haven't changed any settings in the command prompt, etc. It just changed from default insert to default overwrite in CLI apps, but not in the console itself. So, something is different in behavior from 0.4.9 to 1.6.18.

Interesting that without insert mode as default in the command prompt properties, insert mode was still the default in the console prompt, just not in CLI apps.

I also noticed that in 0.4.9 the command prompt respected the setting of the default directory to start in (properties | shortcut | start in) where 1.6.18 always drops me in \windows\system32 no matter what. Probably unrelated, but just something else I noticed.

**Reply by chrisant996** [reply]
Clink doesn't choose the directory.

Something else/more is going on.

> Interesting that without insert mode as default in the command prompt properties, insert mode was still the default in the console prompt, just not in CLI apps.

You said "the console prompt" but you mean "the Clink prompt".  I.e. other non-Clink prompts in the console will have the mode from the Properties page.

Clink sets insert or overwrite based on your .inputrc file config settings.  But Clink only does that during its prompt -- it restores the original mode when launching other programs.  So other programs use the mode from the Properties page. 

That's all normal and expected.  Clink has never been affecting the mode in other programs.

If the behavior changed when you changed to the newer Clink, something else must have changed at the same time.  For example, somehow the mode in Properties changed -- but how did that change?  Clink doesn't change that.

**Reply by jbrown123** [reply]
I'm not sure what happened but whatever it was happened at the same time I updated clink. Both the insert mode and the command prompt autorun / default directory.

At any rate, both are fixed now. Thanks for your help! And thanks for making clink available to all of us.

**Reply by chrisant996** [reply]
How do you start CMD?

Do you always start it by choosing "Clink" from the Start menu? 
If so, then installing a new Clink created a new shortcut file, and will lose any customizations in the previous shortcut file. 

If that's the case, then you might encounter it again each time Clink gets updated, I'm not sure.

And if that's the case, then I would recommend creating your own shortcut that you customize, instead of using the default one installed by Clink.

**Reply by jbrown123** [reply]
I start from a cmd shortcut on my taskbar. That hasn't changed (same shortcut as before). I have clink inject itself via an autorun in the registry.

HKCU\SOFTWARE\Wow6432Node\Microsoft\Command Processor\AutoRun: "C:\Program Files (x86)\clink\clink.bat" inject --autorun --profile ~\clink

It's been a *very* long time since I originally setup 0.4.9 so it's possible I did something like edit the clink.bat and put my stuff in there or some such. Who knows ... I just know it changed when I updated and I don't think I did anything else at the time, but I can't be absolutely positive about that.
